File: Backtracking/0079. Word Search/0079.py
class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        path = set()
        rows = len(board)
        cols = len(board[0])

        def backtrack(i, r, c):
            if i == len(word):
                return True
            # basically if out of bounds OR 
            if (word[i] != board[r][c] or r < 0 or c < 0 or r >= rows or c >= cols or (r, c) in path):
                return False
            
            path.add((r, c))
            res = (backtrack(i + 1, r + 1, c) or backtrack(i + 1, r - 1, c) or backtrack(i + 1, r, c + 1) or
                   backtrack(i + 1, r, c - 1))
            path.remove((r, c))

            return res
        
        for r in range(rows):
            for c in range(cols):
                if backtrack(0, r, c):
                    return True
        
        return False

File: Backtracking/0079. Word Search/0079.cpp
#pragma GCC optimize('0fast', 'unroll-loops')

class Solution {
public:
    bool dfs(int r, int c, int i, string& word, vector<vector<char>>& board, set<pair<int,int>>& path){
        if(i==word.size()){
            return true;
        }

        if(r<0 || r>=board.size() || c<0 || c>=board[0].size() || board[r][c]!=word[i] || path.find({r,c})!=path.end()){
            return false;
        }

        path.insert({r,c});

        bool res=(
            dfs(r+1,c,i+1,word,board,path) || 
            dfs(r-1,c,i+1,word,board,path) || 
            dfs(r,c+1,i+1,word,board,path) || 
            dfs(r,c-1,i+1,word,board,path)
        );

        path.erase({r,c});

        return res;
    }

    bool exist(vector<vector<char>>& board, string word) {
        ios_base::sync_with_stdio(0);
        cout.tie(0);
        cin.tie(0);

        set<pair<int,int>> path;

        for(int r=0; r<board.size(); ++r){
            for(int c=0; c<board[0].size(); ++c){
                if(dfs(r,c,0,word,board,path)){
                    return true;
                }
            }
        }
        return false;
    }
};